google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java

```diff
@@ -0,0 +1,343 @@
+  public void testDmlReturningExecuteQuery() {
+    try (Connection connection = createConnection()) {
+      ResultSet rs = connection.executeQuery(UPDATE_RETURNING_MAP.get(dialect));
+      assertEquals(rs.getColumnCount(), 3);
+      assertTrue(rs.next());
+      assertEquals(rs.getString(1), "ABC");
+      assertTrue(rs.next());
+      assertEquals(rs.getString(1), "ABC");
+      assertTrue(rs.next());
+      assertEquals(rs.getString(1), "ABC");
+      assertFalse(rs.next());
+      assertNotNull(rs.getStats());
+      assertEquals(rs.getStats().getRowCountExact(), 3);
+    }
+  }
+
+  @Test
+  public void testDmlReturningExecuteQueryAsync() {
+    try (Connection connection = createConnection()) {
+      AsyncResultSet rs = connection.executeQueryAsync(UPDATE_RETURNING_MAP.get(dialect));
+      rs.setCallback(
+          Executors.newSingleThreadExecutor(),
+          resultSet -> {
+            try {
+              while (true) {
+                switch (resultSet.tryNext()) {
+                  case OK:
+                    assertEquals(resultSet.getColumnCount(), 3);
+                    assertEquals(resultSet.getString(1), "ABC");
+                    break;
+                  case DONE:
+                    assertNotNull(resultSet.getStats());
+                    assertEquals(resultSet.getStats().getRowCountExact(), 3);
+                    return CallbackResponse.DONE;
+                  case NOT_READY:
+                    return CallbackResponse.CONTINUE;
+                  default:
+                    throw new IllegalStateException();
+                }
+              }
+            } catch (SpannerException e) {
+              System.out.printf("Error in callback: %s%n", e.getMessage());
+              return CallbackResponse.DONE;
+            }
+          });
+    }
+  }
+
+  @Test
+  public void testDmlReturningExecuteUpdate() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      try {
+        connection.executeUpdate(UPDATE_RETURNING_MAP.get(dialect));
+        fail("missing exception");
+      } catch (SpannerException e) {
+        assertEquals(e.getErrorCode(), ErrorCode.FAILED_PRECONDITION);
+      }
+    }
+  }
+
+  @Test
+  public void testDmlReturningExecuteUpdateAsync() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      try {
+        connection.executeUpdateAsync(UPDATE_RETURNING_MAP.get(dialect));
+        fail("missing exception");
+      } catch (SpannerException e) {
+        assertEquals(e.getErrorCode(), ErrorCode.FAILED_PRECONDITION);
+      }
+    }
+  }
+
+  @Test
+  public void testDmlReturningExecuteBatchUpdate() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      final Statement UPDATE_STMT = UPDATE_RETURNING_MAP.get(dialect);
+      long[] counts =
+          connection.executeBatchUpdate(ImmutableList.of(UPDATE_STMT, UPDATE_STMT, UPDATE_STMT));
```

Comment:
If this actually works, then my previous comments regarding `BatchDml` and returning are void. But I also think you should check with the developers of this feature whether this is intentional, as the result of the `RETURNING` clause is lost in these cases.

Reply:
Yes, this feature is intentional, as discussed with the developers of this feature.